File: api/design-patterns/creational/factory/implementations/python/factory.py
"""
Factory Pattern - Python Implementations

Multiple approaches to implement Factory pattern in Python
"""

from abc import ABC, abstractmethod

# ============================================================================
# 1. SIMPLE FACTORY
# ============================================================================
class SimpleFactory:
    @staticmethod
    def create_product(product_type):
        if product_type == 'A':
            return ProductA()
        elif product_type == 'B':
            return ProductB()
        else:
            raise ValueError(f"Unknown product type: {product_type}")

class ProductA:
    def get_name(self):
        return 'Product A'
    
    def do_something(self):
        print('Product A doing something')

class ProductB:
    def get_name(self):
        return 'Product B'
    
    def do_something(self):
        print('Product B doing something')

# ============================================================================
# 2. FACTORY METHOD PATTERN
# ============================================================================
class Product(ABC):
    @abstractmethod
    def operation(self):
        pass

class ConcreteProductA(Product):
    def operation(self):
        return 'Result of ConcreteProductA'

class ConcreteProductB(Product):
    def operation(self):
        return 'Result of ConcreteProductB'

class Creator(ABC):
    @abstractmethod
    def create_product(self):
        pass
    
    def some_operation(self):
        product = self.create_product()
        return f"Creator: {product.operation()}"

class ConcreteCreatorA(Creator):
    def create_product(self):
        return ConcreteProductA()

class ConcreteCreatorB(Creator):
    def create_product(self):
        return ConcreteProductB()

# ============================================================================
# 3. DATABASE CONNECTION FACTORY
# ============================================================================
class DatabaseConnection(ABC):
    @abstractmethod
    def connect(self):
        pass
    
    @abstractmethod
    def query(self, sql):
        pass
    
    @abstractmethod
    def disconnect(self):
        pass

class MySQLConnection(DatabaseConnection):
    def __init__(self, config):
        self.config = config
        self.connected = False
    
    def connect(self):
        print(f"Connecting to MySQL: {self.config['host']}:{self.config['port']}")
        self.connected = True
        return self
    
    def query(self, sql):
        if not self.connected:
            raise Exception("Not connected to database")
        print(f"MySQL Query: {sql}")
        return {"rows": []}
    
    def disconnect(self):
        self.connected = False
        print("MySQL disconnected")

class PostgreSQLConnection(DatabaseConnection):
    def __init__(self, config):
        self.config = config
        self.connected = False
    
    def connect(self):
        print(f"Connecting to PostgreSQL: {self.config['host']}:{self.config['port']}")
        self.connected = True
        return self
    
    def query(self, sql):
        if not self.connected:
            raise Exception("Not connected to database")
        print(f"PostgreSQL Query: {sql}")
        return {"rows": []}
    
    def disconnect(self):
        self.connected = False
        print("PostgreSQL disconnected")

class MongoDBConnection(DatabaseConnection):
    def __init__(self, config):
        self.config = config
        self.connected = False
    
    def connect(self):
        print(f"Connecting to MongoDB: {self.config['host']}:{self.config['port']}")
        self.connected = True
        return self
    
    def query(self, operation):
        if not self.connected:
            raise Exception("Not connected to database")
        print(f"MongoDB Operation: {operation}")
        return {"documents": []}
    
    def disconnect(self):
        self.connected = False
        print("MongoDB disconnected")

class DatabaseFactory:
    @staticmethod
    def create_connection(db_type, config):
        factories = {
            'mysql': MySQLConnection,
            'postgresql': PostgreSQLConnection,
            'mongodb': MongoDBConnection
        }
        
        connection_class = factories.get(db_type.lower())
        if not connection_class:
            raise ValueError(f"Unsupported database type: {db_type}")
        
        return connection_class(config)

# ============================================================================
# 4. PAYMENT PROCESSOR FACTORY
# ============================================================================
class PaymentProcessor(ABC):
    @abstractmethod
    def process_payment(self, amount):
        pass

class CreditCardProcessor(PaymentProcessor):
    def process_payment(self, amount):
        print(f"Processing {amount} via Credit Card")
        return {"success": True, "transaction_id": f"CC-{__import__('time').time()}"}

class PayPalProcessor(PaymentProcessor):
    def process_payment(self, amount):
        print(f"Processing {amount} via PayPal")
        return {"success": True, "transaction_id": f"PP-{__import__('time').time()}"}

class BitcoinProcessor(PaymentProcessor):
    def process_payment(self, amount):
        print(f"Processing {amount} via Bitcoin")
        return {"success": True, "transaction_id": f"BTC-{__import__('time').time()}"}

class PaymentProcessorFactory:
    @staticmethod
    def create_processor(processor_type):
        processors = {
            'creditcard': CreditCardProcessor,
            'paypal': PayPalProcessor,
            'bitcoin': BitcoinProcessor
        }
        
        processor_class = processors.get(processor_type.lower())
        if not processor_class:
            raise ValueError(f"Unsupported payment processor: {processor_type}")
        
        return processor_class()

# ============================================================================
# 5. NOTIFICATION FACTORY
# ============================================================================
class Notification(ABC):
    def __init__(self, recipient, message):
        self.recipient = recipient
        self.message = message
    
    @abstractmethod
    def send(self):
        pass

class EmailNotification(Notification):
    def send(self):
        print(f"Sending email to {self.recipient}: {self.message}")
        return {"success": True, "method": "email"}

class SMSNotification(Notification):
    def send(self):
        print(f"Sending SMS to {self.recipient}: {self.message}")
        return {"success": True, "method": "sms"}

class PushNotification(Notification):
    def send(self):
        print(f"Sending push notification to {self.recipient}: {self.message}")
        return {"success": True, "method": "push"}

class NotificationFactory:
    @staticmethod
    def create_notification(notification_type, recipient, message):
        notifications = {
            'email': EmailNotification,
            'sms': SMSNotification,
            'push': PushNotification
        }
        
        notification_class = notifications.get(notification_type.lower())
        if not notification_class:
            raise ValueError(f"Unsupported notification type: {notification_type}")
        
        return notification_class(recipient, message)

# ============================================================================
# USAGE EXAMPLES
# ============================================================================
if __name__ == "__main__":
    # Simple Factory
    product_a = SimpleFactory.create_product('A')
    product_b = SimpleFactory.create_product('B')
    product_a.do_something()
    product_b.do_something()
    
    # Factory Method
    creator_a = ConcreteCreatorA()
    creator_b = ConcreteCreatorB()
    print(creator_a.some_operation())
    print(creator_b.some_operation())
    
    # Database Factory
    mysql = DatabaseFactory.create_connection('mysql', {'host': 'localhost', 'port': 3306})
    mysql.connect().query('SELECT * FROM users')
    
    postgres = DatabaseFactory.create_connection('postgresql', {'host': 'localhost', 'port': 5432})
    postgres.connect().query('SELECT * FROM users')
    
    # Payment Factory
    credit_card = PaymentProcessorFactory.create_processor('creditcard')
    credit_card.process_payment(100)
    
    paypal = PaymentProcessorFactory.create_processor('paypal')
    paypal.process_payment(200)
    
    # Notification Factory
    email = NotificationFactory.create_notification('email', 'user@example.com', 'Hello!')
    email.send()
    
    sms = NotificationFactory.create_notification('sms', '+1234567890', 'Hello!')
    sms.send()

File: api/design-patterns/creational/factory/implementations/java/Factory.java
/**
 * Factory Pattern - Java Implementations
 * 
 * Multiple approaches to implement Factory pattern in Java
 */

// ============================================================================
// 1. SIMPLE FACTORY
// ============================================================================
class SimpleFactory {
    public static Product createProduct(String type) {
        switch (type) {
            case "A":
                return new ProductA();
            case "B":
                return new ProductB();
            default:
                throw new IllegalArgumentException("Unknown product type: " + type);
        }
    }
}

interface Product {
    String getName();
    void doSomething();
}

class ProductA implements Product {
    @Override
    public String getName() {
        return "Product A";
    }
    
    @Override
    public void doSomething() {
        System.out.println("Product A doing something");
    }
}

class ProductB implements Product {
    @Override
    public String getName() {
        return "Product B";
    }
    
    @Override
    public void doSomething() {
        System.out.println("Product B doing something");
    }
}

// ============================================================================
// 2. FACTORY METHOD PATTERN
// ============================================================================
interface ProductInterface {
    String operation();
}

class ConcreteProductA implements ProductInterface {
    @Override
    public String operation() {
        return "Result of ConcreteProductA";
    }
}

class ConcreteProductB implements ProductInterface {
    @Override
    public String operation() {
        return "Result of ConcreteProductB";
    }
}

abstract class Creator {
    public abstract ProductInterface createProduct();
    
    public String someOperation() {
        ProductInterface product = createProduct();
        return "Creator: " + product.operation();
    }
}

class ConcreteCreatorA extends Creator {
    @Override
    public ProductInterface createProduct() {
        return new ConcreteProductA();
    }
}

class ConcreteCreatorB extends Creator {
    @Override
    public ProductInterface createProduct() {
        return new ConcreteProductB();
    }
}

// ============================================================================
// 3. DATABASE CONNECTION FACTORY
// ============================================================================
interface DatabaseConnection {
    DatabaseConnection connect();
    void query(String sql);
    void disconnect();
}

class MySQLConnection implements DatabaseConnection {
    private String host;
    private int port;
    private boolean connected = false;
    
    public MySQLConnection(String host, int port) {
        this.host = host;
        this.port = port;
    }
    
    @Override
    public DatabaseConnection connect() {
        System.out.println("Connecting to MySQL: " + host + ":" + port);
        this.connected = true;
        return this;
    }
    
    @Override
    public void query(String sql) {
        if (!connected) {
            throw new RuntimeException("Not connected to database");
        }
        System.out.println("MySQL Query: " + sql);
    }
    
    @Override
    public void disconnect() {
        this.connected = false;
        System.out.println("MySQL disconnected");
    }
}

class PostgreSQLConnection implements DatabaseConnection {
    private String host;
    private int port;
    private boolean connected = false;
    
    public PostgreSQLConnection(String host, int port) {
        this.host = host;
        this.port = port;
    }
    
    @Override
    public DatabaseConnection connect() {
        System.out.println("Connecting to PostgreSQL: " + host + ":" + port);
        this.connected = true;
        return this;
    }
    
    @Override
    public void query(String sql) {
        if (!connected) {
            throw new RuntimeException("Not connected to database");
        }
        System.out.println("PostgreSQL Query: " + sql);
    }
    
    @Override
    public void disconnect() {
        this.connected = false;
        System.out.println("PostgreSQL disconnected");
    }
}

class MongoDBConnection implements DatabaseConnection {
    private String host;
    private int port;
    private boolean connected = false;
    
    public MongoDBConnection(String host, int port) {
        this.host = host;
        this.port = port;
    }
    
    @Override
    public DatabaseConnection connect() {
        System.out.println("Connecting to MongoDB: " + host + ":" + port);
        this.connected = true;
        return this;
    }
    
    @Override
    public void query(String operation) {
        if (!connected) {
            throw new RuntimeException("Not connected to database");
        }
        System.out.println("MongoDB Operation: " + operation);
    }
    
    @Override
    public void disconnect() {
        this.connected = false;
        System.out.println("MongoDB disconnected");
    }
}

class DatabaseFactory {
    public static DatabaseConnection createConnection(String type, String host, int port) {
        switch (type.toLowerCase()) {
            case "mysql":
                return new MySQLConnection(host, port);
            case "postgresql":
                return new PostgreSQLConnection(host, port);
            case "mongodb":
                return new MongoDBConnection(host, port);
            default:
                throw new IllegalArgumentException("Unsupported database type: " + type);
        }
    }
}

// ============================================================================
// 4. PAYMENT PROCESSOR FACTORY
// ============================================================================
interface PaymentProcessor {
    PaymentResult processPayment(double amount);
}

class PaymentResult {
    private boolean success;
    private String transactionId;
    
    public PaymentResult(boolean success, String transactionId) {
        this.success = success;
        this.transactionId = transactionId;
    }
    
    public boolean isSuccess() {
        return success;
    }
    
    public String getTransactionId() {
        return transactionId;
    }
}

class CreditCardProcessor implements PaymentProcessor {
    @Override
    public PaymentResult processPayment(double amount) {
        System.out.println("Processing " + amount + " via Credit Card");
        return new PaymentResult(true, "CC-" + System.currentTimeMillis());
    }
}

class PayPalProcessor implements PaymentProcessor {
    @Override
    public PaymentResult processPayment(double amount) {
        System.out.println("Processing " + amount + " via PayPal");
        return new PaymentResult(true, "PP-" + System.currentTimeMillis());
    }
}

class BitcoinProcessor implements PaymentProcessor {
    @Override
    public PaymentResult processPayment(double amount) {
        System.out.println("Processing " + amount + " via Bitcoin");
        return new PaymentResult(true, "BTC-" + System.currentTimeMillis());
    }
}

class PaymentProcessorFactory {
    public static PaymentProcessor createProcessor(String type) {
        switch (type.toLowerCase()) {
            case "creditcard":
                return new CreditCardProcessor();
            case "paypal":
                return new PayPalProcessor();
            case "bitcoin":
                return new BitcoinProcessor();
            default:
                throw new IllegalArgumentException("Unsupported payment processor: " + type);
        }
    }
}

// ============================================================================
// 5. NOTIFICATION FACTORY
// ============================================================================
interface Notification {
    NotificationResult send();
}

class NotificationResult {
    private boolean success;
    private String method;
    
    public NotificationResult(boolean success, String method) {
        this.success = success;
        this.method = method;
    }
    
    public boolean isSuccess() {
        return success;
    }
    
    public String getMethod() {
        return method;
    }
}

class EmailNotification implements Notification {
    private String recipient;
    private String message;
    
    public EmailNotification(String recipient, String message) {
        this.recipient = recipient;
        this.message = message;
    }
    
    @Override
    public NotificationResult send() {
        System.out.println("Sending email to " + recipient + ": " + message);
        return new NotificationResult(true, "email");
    }
}

class SMSNotification implements Notification {
    private String recipient;
    private String message;
    
    public SMSNotification(String recipient, String message) {
        this.recipient = recipient;
        this.message = message;
    }
    
    @Override
    public NotificationResult send() {
        System.out.println("Sending SMS to " + recipient + ": " + message);
        return new NotificationResult(true, "sms");
    }
}

class PushNotification implements Notification {
    private String recipient;
    private String message;
    
    public PushNotification(String recipient, String message) {
        this.recipient = recipient;
        this.message = message;
    }
    
    @Override
    public NotificationResult send() {
        System.out.println("Sending push notification to " + recipient + ": " + message);
        return new NotificationResult(true, "push");
    }
}

class NotificationFactory {
    public static Notification createNotification(String type, String recipient, String message) {
        switch (type.toLowerCase()) {
            case "email":
                return new EmailNotification(recipient, message);
            case "sms":
                return new SMSNotification(recipient, message);
            case "push":
                return new PushNotification(recipient, message);
            default:
                throw new IllegalArgumentException("Unsupported notification type: " + type);
        }
    }
}

// ============================================================================
// USAGE EXAMPLES
// ============================================================================
public class Factory {
    public static void main(String[] args) {
        // Simple Factory
        Product productA = SimpleFactory.createProduct("A");
        Product productB = SimpleFactory.createProduct("B");
        productA.doSomething();
        productB.doSomething();
        
        // Factory Method
        Creator creatorA = new ConcreteCreatorA();
        Creator creatorB = new ConcreteCreatorB();
        System.out.println(creatorA.someOperation());
        System.out.println(creatorB.someOperation());
        
        // Database Factory
        DatabaseConnection mysql = DatabaseFactory.createConnection("mysql", "localhost", 3306);
        mysql.connect().query("SELECT * FROM users");
        
        DatabaseConnection postgres = DatabaseFactory.createConnection("postgresql", "localhost", 5432);
        postgres.connect().query("SELECT * FROM users");
        
        // Payment Factory
        PaymentProcessor creditCard = PaymentProcessorFactory.createProcessor("creditcard");
        creditCard.processPayment(100);
        
        PaymentProcessor paypal = PaymentProcessorFactory.createProcessor("paypal");
        paypal.processPayment(200);
        
        // Notification Factory
        Notification email = NotificationFactory.createNotification("email", "user@example.com", "Hello!");
        email.send();
        
        Notification sms = NotificationFactory.createNotification("sms", "+1234567890", "Hello!");
        sms.send();
    }
}
